lib/safari.js

```diff
@@ -1,53 +1,173 @@
+  get isRunning () {
+    return !!(this.proc?.isRunning);
+  }
+
+  async init () {
+    await SAFARI_SERVER_GUARD(async () => {
+      if (this.isRunning) {
+        return;
+      }
+
+      const [startPort, endPort] = SAFARI_PORT_RANGE;
+      try {
+        this.port = await findAPortNotInUse(startPort, endPort);
+      } catch (e) {
+        throw new Error(
+          `Cannot find any free port in range ${startPort}..${endPort}}. ` +
+          `Double check the processes that are locking ports within this range and terminate ` +
+          `these which are not needed anymore`);
+      }
+
+      let safariBin;
+      try {
+        safariBin = await fs.which(SD_BINARY);
+      } catch (e) {
+        throw new Error(`${SD_BINARY} binary cannot be found in PATH. ` +
+          `Please make sure it is present on your system`);
+      }
+      this.proc = new SubProcess(safariBin, ['-p', this.port, '--diagnose']);
+      this.proc.on('output', (stdout, stderr) => {
+        const line = stdout || stderr;
+        log.debug(`[${SD_BINARY}] ${line}`);
+      });
+      this.proc.on('exit', (code, signal) => {
+        log.info(`${SD_BINARY} has exited with code ${code}, signal ${signal}`);
+      });
+      log.info(`Starting ${SD_BINARY} on port ${this.port}`);
+      await this.proc.start(0);
+    });
+  }
+
+  async kill () {
+    if (this.isRunning) {
+      try {
+        await this.proc.stop('SIGKILL');
+      } catch (ign) {}
+    }
+  }
+}
 
-    this.proc.on('stream-line', log.debug.bind(log));
-    this.proc.on('stop', this.onProcStop.bind(this));
-    this.proc.on('die', this.onProcDie.bind(this));
+// Single server process per Appium instance
+const SAFARI_DRIVER_PROCESS = new SafariDriverProcess();
+process.once('exit', () => {
+  if (SAFARI_DRIVER_PROCESS.isRunning) {
+    try {
+      execSync(`kill ${SAFARI_DRIVER_PROCESS.proc.pid}`);
+    } catch (ign) {}
+  }
+});
 
-    log.info('Starting safaridriver');
-    await this.proc.start(0);
+class SafariDriverServer {
+  constructor () {
+    this.proxy = null;
   }
 
-  onProcStop (code, signal) {
-    log.info(`Safaridriver exited cleanly with code ${code} and signal ${signal}`);
+  get isRunning () {
+    return !!(SAFARI_DRIVER_PROCESS.isRunning);
   }
 
-  onProcDie (code, signal) {
-    const errMsg = `Safaridriver exited unexpectedly with code ${code} ` +
-                   `and signal ${signal}`;
-    log.error(errMsg);
-    if (this.unexpectedExitHandler) {
-      this.unexpectedExitHandler(new Error(errMsg));
+  async start (caps) {
+    await SAFARI_DRIVER_PROCESS.init();
+
+    this.proxy = new SafariProxy({
+      server: '127.0.0.1',
+      port: SAFARI_DRIVER_PROCESS.port,
+      base: '',
+      keepAlive: true,
+    });
+    this.proxy.didProcessExit = false;
+    SAFARI_DRIVER_PROCESS.proc.on('exit', () => {
+      this.proxy.didProcessExit = true;
+    });
+
+    try {
+      await waitForCondition(async () => {
```

Comment:
we could put this start wait logic in the `this.proc.start` section above, i believe teen process has a way to provide a custom start awaiter? not sure what makes  more sense

Reply:
I tried that a quite while ago and it did not work as expected and also looked too complicated. So I decided to use more explicit and obvious solution. Also, it makes sense to check the status before each session, because we use single server instance there; just to make sure everything is fine before session startup